Sprint 7 solution in file manager

GritsED opened the pull request:
Реализовал 7 спринт

GritsED committed:
Implement 7 sprint

Add new class FileBackedTaskManager:
Add methods save, loadFromFile, fromString

Add new enum Type
Add 2 exception classes

GritsED committed:
Perform tests to verify the functionality

GritsED committed:
Some fixes

Add type, getType
Override toString

vlolad commented on src/model/Task.java:
```diff
@@ -8,6 +8,8 @@ public class Task {
     protected String description;
     protected TaskStatus status;
 
+    protected final Type type = Type.TASK;
```
У всех наследников дублируется поле type, это не очень хорошо
Есть вариант проще - оставь геттер getType() и в нем возвращай требуемый тип задачи. В наследниках getType() можно переопределить, чтобы они возвращали корректный тип

Вариант сложнее, но допустимый - убрать с поля final и во всех конструкторах наследника менять type на соответствующий классу

vlolad commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,257 @@
+        }
+    }
+
+    public static Task fromString(String value) {
```
Лучше всего конвертацию в/из строки вынести в отдельный класс, например CSVTaskConverter. В нем будет два метода на конвертацию туда-обратно (можно больше), и они будут вызываться в менеджере

Обосновано тем, что сериализация бывает в разных форматах представлена, и если потребуется преобразовывать не в CSV - придется лезть в код менеджера. А так можно будет заменить реализацию конвертера, например. Преобразование задач и парсинг строк - не зона ответственности менеджера

vlolad commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,257 @@
+        TaskStatus status;
+        String description = contents[4];
+
+        switch (contents[3]) {
```
Лучше сделать отдельную переменную и её прокидывать в switch. Повысится читаемость кода. contents[3] - непонятно

vlolad commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,257 @@
+                status = TaskStatus.NEW;
+        }
+
+        switch (contents[1]) {
```
Аналогично тому, что написано выше

vlolad commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,257 @@
+            case "IN_PROGRESS":
+                status = TaskStatus.IN_PROGRESS;
+                break;
+            case "Done":
```
Другой регистр, надо поправить

vlolad commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,257 @@
+                Task task = fromString(line);
+
+                if (maxId > taskManager.id) {
+                    taskManager.id = maxId + 1;
```
Хорошо, что учел обновление идентификаторов

vlolad commented on src/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,257 @@
+    public Task addNewTask(Task newTask) {
+        Task addedTask = super.addNewTask(newTask);
+
+        try {
```
Во всех методах обертка try - catch, которая проверяет ошибку, прилетающую из save(). Проще её сразу в save и обрабатывать, так не будет лишнего дублирования кода
ManagerSaveException и его дубликат - хорошая практика, но в текущем варианте лишнее

Если хочется выбрасывать ошибку в save() и обрабатывать её - можно разделить save() на два метода. Первый будет содержать логику текущего save(), а второй - вызывать первый, но оборачивать его в try-catch

vlolad commented on src/exceptions/ManagerSaveException.java:
```diff
@@ -0,0 +1,8 @@
+package exceptions;
+
+public class ManagerSaveException extends RuntimeException {
```
Одинаковые ошибки ManagerSaveException и ManagerLoadException, можно оставить что-то одно (по логике - лучше первое)

vlolad commented on src/service/Managers.java:
```diff
@@ -1,12 +1,22 @@
+    }
+
     public static TaskManager getDefault() {
         return new InMemoryTaskManager();
```
getDefault() теперь возвращает FileBackedTaskManager, т.к. основная реализация сменилась. Не знаю, почему в ТЗ это не отмечают :с
Файл можно вынести в константу на уровне Managers или сделать конструктор без параметров (дополнительный) в самом классе, чтобы можно было вызывать дефолтную реализацию в дефолтном файле

vlolad commented on test/service/AbstractTaskManagerTest.java:
```diff
@@ -0,0 +1,361 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+abstract class AbstractTaskManagerTest<T extends TaskManager> {
```
Отлично, что реализовал общие тесты через абстрактный класс

vlolad reviewed (changes requested):
Привет!

Логика работы новой фичи верная, но по коду есть пару замечаний в части оптимизации и декомпозиции. Надо немного доработать
Вроде в комментариях всё подробно описал, жду на повторную проверку. Хороших выходных!

GritsED committed:
Some fixes
Added class TaskConverter
getDefault() returns FileBackedTaskManager
Deleted ManagerLoadException
Delete duplicate code
Deleted "try...catch" from the methods of modification

GritsED committed:
Some fixes
Added class TaskConverter
getDefault() returns FileBackedTaskManager
Deleted ManagerLoadException
Delete duplicate code
Deleted "try...catch" from the methods of modification

vlolad commented on src/model/Epic.java:
```diff
@@ -15,6 +15,15 @@ public Epic(String name, String description) {
+
+    @Override
+    public Type getType() {
+        return type = Type.EPIC;
```
Можно просто return Type.EPIC, но и так сойдет

vlolad commented on src/service/TaskConverter.java:
```diff
@@ -0,0 +1,52 @@
+import model.Task;
+import model.TaskStatus;
+
+public class TaskConverter {
```
Отлично!

vlolad reviewed (approved):
Привет!

Хорошая работа по рефакторингу, принято с:
